engines/director/lingo/xlibs/mmovie.cpp

```diff
@@ -126,32 +250,411 @@ void MMovieXObj::close(ObjectType type) {
+	g_lingo->push(result);
+}
+
+void MMovieXObj::m_getMovieRate(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_getMovieRate", nargs);
+	if (nargs != 0) {
+		g_lingo->dropStack(nargs);
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	g_lingo->push(Datum(me->_rate));
+}
+
+void MMovieXObj::m_setMovieRate(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_setMovieRate", nargs);
+	if (nargs != 1) {
+		warning("MMovieXObj::m_setMovieRate: expecting 4 arguments");
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Datum(0));
+		return;
+	}
+	MMovieXObject *me = static_cast<MMovieXObject *>(g_lingo->_state->me.u.obj);
+	me->_rate = g_lingo->pop().asInt();
+	g_lingo->push(Datum(me->_rate));
+}
+
 XOBJSTUB(MMovieXObj::m_flushEvents, 0)
 XOBJSTUB(MMovieXObj::m_invalidateRect, 0)
-XOBJSTUB(MMovieXObj::m_readFile, "")
-XOBJSTUB(MMovieXObj::m_writeFile, "")
+
+void MMovieXObj::m_readFile(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_readFile", nargs);
+	if (nargs != 2) {
+		warning("MMovieXObj::m_readFile(): expecting 2 argument");
+	}
+	Common::SaveFileManager *saves = g_system->getSavefileManager();
+	bool scramble = g_lingo->pop().asInt() != 0;
+	Common::String origPath = g_lingo->pop().asString();
+	Common::String path = origPath;
+
+	Common::String prefix = g_director->getTargetName() + '-';
+	Common::String result;
+	if (origPath.empty()) {
+		Common::String mask = prefix + "*.txt";
+
+		GUI::FileBrowserDialog browser(nullptr, "txt", GUI::kFBModeLoad, mask.c_str());
+		if (browser.runModal() <= 0) {
+			debugC(5, kDebugXObj, "MMovieXObj::m_readFile(): read cancelled by modal");
+			g_lingo->push(result);
+			return;
+		}
+		path = browser.getResult();
+
+	} else {
+		path = lastPathComponent(origPath, g_director->_dirSeparator);
+		if (path.hasSuffixIgnoreCase(".txt"))
+			path += ".txt";
+	}
+	if (!path.hasPrefixIgnoreCase(prefix)) {
+		path = prefix + path;
+	}
+
+	Common::SeekableReadStream *stream = saves->openForLoading(path);
+	if (stream) {
+		debugC(5, kDebugXObj, "MMovieXObj::m_readFile(): opening file %s as %s from the saves dir", origPath.c_str(), path.c_str());
+	} else {
+		Common::File *f = new Common::File;
+		Common::Path location = findPath(origPath);
+		if (!location.empty() && f->open(location)) {
+			debugC(5, kDebugXObj, "MMovieXObj::m_readFile(): opening file %s from the game dir", origPath.c_str());
+			stream = (Common::SeekableReadStream *)f;
+		} else {
+			delete f;
+		}
+	}
+
+	if (stream) {
+		while (!stream->eos() && !stream->err()) {
+			byte ch = stream->readByte();
+			if (scramble) // remove unbreakable encryption
+				ch ^= 0xa5;
+			result += ch;
+		}
+		delete stream;
+	} else {
+		warning("MMovieXObj::m_readFile(): file %s not found", origPath.c_str());
+	}
+
+	g_lingo->push(result);
+}
+
+void MMovieXObj::m_writeFile(int nargs) {
+	g_lingo->printArgs("MMovieXObj::m_writeFile", nargs);
+	if (nargs != 3) {
+		warning("MMovieXObj::m_writeFile(): expecting 3 arguments");
```

Comment:
Should it exit? Because if less or more, you'll mess up the stack

Reply:
Fixed